### ISSUE TITLE
support NDList with more than one item

### ISSUE THREAD

**Issue by patins1** [deepjavalibrary/djl#1499]
## Description

While technically, an NDList instance can have more than one item, practically it is not well supported by DJL,
a good example is the LambdaBlock::singleton block (or the Linear block) that throws an exception when there is more than one.

Will this change the current api? How?

I think there needs to be a configurable flag `multiInput` in `AbstractBlock `that determines if e.g. `forward()` is to be applied on every item in the NDList producing a same-length NDList (`true`) or if the old behavior is to be applied (`false`). 
In the former case, e.g. a Linear block would hold one parameter array for each item in the input list.

Who will benefit from this enhancement?

As my current workaround of routing each item of the NDList into its own block, I would benefit if this feature can utilize parallelism on the gpu (my current performance stats are linear inverse to the number of items in the NDList..). If parallelization is not possible , then this feature request is not really useful, but I cannot answer this question as I am not familiar with PyTorch or GPU architecture.

**Comment by zachgk** [comment]
I think there might be a bit of confusion with using NDArrays. So, the typical `LambdaBlock` is `NDList -> NDList` and it what you want. You can create it with `new LambdaBlock(...)`. The `LambdaBlock.singleton(...)` is a helper we made to handle the case of when your NDLists only have a single item.

Now, as to the larger question of how to apply a block to multiple inputs. Let's say our input is an image which has shape `chw` or channels (RGB) * height * width. Then, we have `n` images. The way you are thinking to represent this is `NDList(chw, chw, chw, ...)` where you then need this kind of feature. Instead, most of DJL represents it as a single NDArray of shape `nchw` or `NDList(nchw)`. Then, most blocks will automatically apply the changes to every image in your "batch" NDArray.

**Comment by patins1** [comment]
Thanks for your answer, I guess it wont help me with my problem. `LambdaBlock.singleton()` might be a bad example, it is actually the `Linear `block I am aiming for. Inputting your conceptual `nchw` into a `Linear ` block doesn't work as  - depending of how you arrange the original `chw` items , above each other or next to each other into `nchw `-  the one parameter array for the `Linear `block melts together the pixel information of all images in one way or the other. The logic I am really looking for must be equivalent to apply a separate `Linear `block to each original picture, but utilizing parallelism like in your approach .

**Comment by zachgk** [comment]
The problem you are having with linear block is that it accepts 1-dimensional feature vector input, while an image is 3-dimensional (including channels). One thing you can do is reshape your image `Shape(n, c, h, w)` and flatten it to `Shape(n, c*h*w)` and then apply the linear block.

The other option is to apply the `Conv2D` operation which works on local patches of your image, usually 3 pixel by 3 pixel squares. Most computer vision models use this. If you do a `1x1` convolution, that is equivalent to applying a `Linear` transformation to each pixel in your image independently.

Finally, you may be interested in looking into the [Dive into Deep Learning book](https://d2l.djl.ai/). It has some chapters about linear networks and convolutional networks that may help.

**Comment by patins1** [comment]
I already flatten the input features using `Blocks.batchFlattenBlock()` , but this doesn't help with my problem. It is mathematical not possible to express a set of linear transformations `A_1*w_1,  A_2*w_2, ..., A_n*w_n`  (I do not use a bias) as one big transformation `A*w`, therefor my enhancement request. Giving a simple example with flattening applied, for
```

(x1,x2)*(w1,w2) = x1*w1+x2*w2 = y1
(x3,x4)*(w3,w4) = x3*w3+x4*w4 = y2
```

you cannot extract y1 and y2 putting everything into one NDArray as in

`(x1,x2,x3,x4) * (y1,y2,y3,y4) = x1*w1+x2*w2+x3*w3+x4*w4 = y1+y2`

because y1 and y2 are unfortunately added together, i need them separately.
I haven't looked into `Conv2D `as i dont deal with image data, but do you think it can provide that?

**Comment by zachgk** [comment]
In your example, are x1 and x2 vectors or scalars? If they are scalars then you need two linear blocks because each corresponds to a separate weight. If they are vectors, you would need four. Or, you can also use the static `Linear.linear(x, w)` if that makes it easier.

I feel like we may be in a bit of an [XY problem](https://xyproblem.info/). Can you describe the particular operation you are trying to accomplish? What format is your data in (array shape(s))? If you have code or pseudo-code, can you share it?

**Comment by patins1** [comment]
In my example, x1 and x2 are scalars. as you already said - you would need separate weights, in my workaround (for the missing enhancement), i use separate blocks (that do not only contain a Linear block each), but I realized that doesnt scale well, i.e. those blocks are not executed in parallel.  I am not sure if machine learning hardware supports parallelization of blocks in general or not.

Regarding the XY problem, what I generally want to achieve is to train N models, each model has the same underlying neural network but using different features/labels each. To execute them not sequentially which takes a LONG time, my initial idea was to solve this in one overarching neural network, by summing up the losses of each original network in an overarching loss function (this works as each original loss >=0), and , this was the idea, by processing the features in separate items to the input NDList.

**Comment by zachgk** [comment]
Here is a question. Rather than having N models, have you considered making 1 model that has all of the N output features? You can take the full output, split it into N smaller output vectors, find a separate loss for each N, and then combine the losses. This way, if there is anything which could be shared between the features in your model, it may reduce duplication of work. It may also be worth making this single model somewhat larger than the original ones as well, but since you are reducing the work by a factor of N it should still be much faster.

Otherwise, it sounds like your N models are fully independent. Even if you combine the loss, if there are no shared parameters between the models then it will not work differently from just training N models separately. In that case, why not just train the models separately? You could even just different GPUs or hardware to train each of the models.

**Comment by patins1** [comment]
The approach you are outlining is probably what i was trying to do already, it can be reduced to the problem that you cannot partition a Linear block , unless I am taught differently. Keep in mind that I have N completely independent model. About 500 hundred of them. And there are absolutely no shared parameters. Just training N models separately will cost N times training a single model (divided by maybe 2 GPUs), but the whole point of my enhancement request is to do that in parallel. I am just not sure how to wire the blocks, and I am curious if PyTorch could do the job, but i dont want to learn Python :(

**Comment by patins1** [comment]
I think the crucial point is  "If you have two independent  (NDArray) operations, they would be run in sequence." Cant they run on different threads? Are GPUs not threaded?

**Comment by zachgk** [comment]
GPUs are threaded, but in a different way than CPUs. A CPU is N cores and the cores are fully independent. They can each work independently. GPUs have many more cores, but try to do the same instruction across all cores at the same time. They work so well with NDArrays because there are a lot of tasks where the same operation is happening for each NDArray element and the cores can each work on a separate element

**Comment by patins1** [comment]
Enhancement is covered by the new LinearCollection block type: it does not operate on multiple items in a NDList according to the enhancement description, but it splits a single NDArray by having (at least one) split dimension added, thus having multiple linear transformations parallelized and executed by LinearCollection.
